File: Python/omniPy/Stats/cov_matrix.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import numpy as np
import sys, warnings

def cov_matrix( x , y = None , rowvar = False ) -> 'Covariance between each column in a matrix to all others':
    #000.   Info.
    '''
#---------------------------------------------------------------------------------------------------------------------------------------#
#100.   Introduction.                                                                                                                   #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |This function is intended to calculate the Covariance Matrix for each column in the matrix to all other columns                    #
#   |This function calculates much faster than the internal function [np.cov()]                                                         #
#   |Quote: https://blog.csdn.net/lph188/article/details/84501481                                                                       #
#   |Quick solution is as below:                                                                                                        #
#   |n <- nrow(x)                                                                                                                       #
#   |#Below function [diag(1,100000)] will consume 74.5GB RAM! Thus it cannot be applied to mass calculation.                           #
#   |mx <- diag(1,n) - matrix(1,n,n) / n                                                                                                #
#   |cov <- t(x) %*% mx %*% x)/(n-1)                                                                                                    #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |Difference between this function, [np.cov] and R:                                                                                  #
#   |[1] This function is able to be applied to two different matrices, while [np.cov] can only be applied to a single matrix           #
#   |[2] This function is slightly slower than [np.cov] on large matrix                                                                 #
#   |[3] Both functions in Python are twice faster than R with much less CPU effort (only when Rcpp is applied for calculation based on #
#   |     C++ optmization)                                                                                                              #
#---------------------------------------------------------------------------------------------------------------------------------------#
#200.   Glossary.                                                                                                                       #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Parameters.                                                                                                                 #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |x,y        :   The input matrices for which the calculation is to be taken upon the columns                                        #
#   |rowvar     :   Whether the requested calculation is applied to each row to all others (Compatible to [numpy])                      #
#   |               [False]<Default> Calculate the distance between each column in [x] to that in [y]                                   #
#   |               [True]           Calculate the distance between each row in [x] to that in [y]                                      #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |900.   Return Values by position.                                                                                                  #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |[matrix]   :   The [K*M] matrix, where [K] is equal to the number of columns of [x], while [M] is the number of columns of [y]     #
#   |               Each [k,m] represents the covariance of [k]th column in [x] to [m]th column in [y]                                  #
#---------------------------------------------------------------------------------------------------------------------------------------#
#300.   Update log.                                                                                                                     #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   | Date |    20200606        | Version | 1.00        | Updater/Creator | Lu Robin Bin                                                #
#   |______|____________________|_________|_____________|_________________|_____________________________________________________________#
#   | Log  |Version 1                                                                                                                   #
#   |______|____________________________________________________________________________________________________________________________#
#---------------------------------------------------------------------------------------------------------------------------------------#
#400.   User Manual.                                                                                                                    #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |See the [Full Test Program] section                                                                                                #
#---------------------------------------------------------------------------------------------------------------------------------------#
#500.   Dependent Facilities.                                                                                                           #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Dependent packages                                                                                                          #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |   |numpy, sys, warnings                                                                                                           #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |300.   Dependent user-defined functions                                                                                            #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------------------------------------------#
    '''

    #001.   Import necessary functions for processing.

    #010.   Check parameters.
    #011.   Prepare log text.
    #python 动态获取当前运行的类名和函数名的方法: https://www.cnblogs.com/paranoia/p/6196859.html
    LfuncName : str = sys._getframe().f_code.co_name

    #012.   Handle the parameter buffer.
    if not isinstance( x , ( np.ndarray , np.matrix ) ):
        raise TypeError( '[' + LfuncName +  '][x] should be of the type [np.matrix]! Type of input value is [{0}]'.format( type(x) ) )
    chkNaN_x = min( x[np.isnan(x)].shape ) != 0
    chkNaN_y = False
    if y is not None:
        if not isinstance( y , ( np.ndarray , np.matrix ) ):
            raise TypeError( '[' + LfuncName +  '][y] should be of the type [np.matrix]! Type of input value is [{0}]'.format( type(y) ) )
        chkNaN_y = min( y[np.isnan(y)].shape ) != 0
    chkNaN = chkNaN_x | chkNaN_y
    if chkNaN:
        warnings.warn( '[' + LfuncName +  ']NaN values are found, [np.nanmean] is used instead! Result may be unexpected!' )
        f_mean = np.nanmean
    else:
        f_mean = np.mean

    #013.   Define the local environment.

    #050.   Transpose [x] if it is requested for calculation based on [row]s.
    if isinstance( x , ( np.ndarray ) ): x = np.asmatrix(x)
    if rowvar: x = x.T

    #100.   Reshape [x].
    x = x.astype(np.float64)
    n = x.shape[0]
    x -= f_mean(x, axis = 0)

    #300.   Further handle [y] if it is provided.
    if y is None: y = x
    elif y is x: pass
    else:
        #050.   Transpose [x] if it is requested for calculation based on [row]s.
        if isinstance( y , ( np.ndarray ) ): y = np.asmatrix(y)
        if rowvar: y = y.T

        #100.   Reshape [y].
        y = y.astype(np.float64)
        y -= f_mean(y, axis = 0)

    #900.   Output.
    return( np.dot(x.T, y) / (n - 1) )
#End cov_matrix

'''
#-Notes- -Begin-
#Full Test Program[1]:
if __name__=='__main__':
    #010.   Create envionment.
    import time
    import sys
    dir_omniPy : str = r'D:\Python\ '.strip()
    if dir_omniPy not in sys.path:
        sys.path.append( dir_omniPy )
    from omniPy.Stats import cov_matrix
    print(cov_matrix.__doc__)

    #[AMD FX-6300 6Core 3.5G]

    #100.   Create the testing dataset.
    x1 = np.random.randn(10,5)
    y1 = np.random.randn(10,5)
    isinstance(x1,( np.ndarray ))

    #200.   Calculate the covariance matrix of a single matrix.
    t0 = time.time()
    c1 = cov_matrix(x1)
    print(time.time() - t0)

    t0 = time.time()
    c2 = np.cov(x1, rowvar = False)
    print(time.time() - t0)

    np.allclose(c1,c2)

    #500.   Create the large matrices.
    x2 = np.random.randn(100000,1000)
    y2 = np.random.randn(100000,500)

    #500. Calculation upon large matrix
    t0 = time.time()
    c3 = cov_matrix(x2)
    print(time.time() - t0)
    #5.08s

    t0 = time.time()
    c4 = np.cov(x2, rowvar = False)
    print(time.time() - t0)
    #4.85s

    np.allclose(c3,c4)

    #600. Calculation upon two different matrices.
    t0 = time.time()
    c_xy1 = cov_matrix(x2, y2)
    print(time.time() - t0)
    #4.82s
#-Notes- -End-
'''

File: R/omniR/Stats/cov_matrix.r
#---------------------------------------------------------------------------------------------------------------------------------------#
#100.   Introduction.                                                                                                                   #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |This function is intended to calculate the Covariance Matrix for each column in the matrix to all other columns                    #
#   |This function calculates much faster than the internal function [cov()]                                                            #
#   |Quote: https://blog.csdn.net/lph188/article/details/84501481                                                                       #
#   |Quick solution is as below:                                                                                                        #
#   |n <- nrow(x)                                                                                                                       #
#   |#Below function [diag(1,100000)] will consume 74.5GB RAM! Thus it cannot be applied to mass calculation.                           #
#   |mx <- diag(1,n) - matrix(1,n,n) / n                                                                                                #
#   |cov <- t(x) %*% mx %*% x)/(n-1)                                                                                                    #
#---------------------------------------------------------------------------------------------------------------------------------------#
#200.   Glossary.                                                                                                                       #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Parameters.                                                                                                                 #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |x,y        :   The input matrices for which the calculation is to be taken upon the columns                                        #
#   |dim        :   On which dimension is the calculation requested                                                                     #
#   |               [row] Calculate the distance between each row in [x] to that in [y]                                                 #
#   |               [col] Calculate the distance between each column in [x] to that in [y]                                              #
#   |               [Default] [col] (covariance is usually calculated for models that are based upon dimensions)                        #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |900.   Return Values by position.                                                                                                  #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |[matrix]   :   The [K*M] matrix, where [K] is equal to the number of columns of [x], while [M] is the number of columns of [y]     #
#   |               Each [k,m] represents the covariance of [k]th column in [x] to [m]th column in [y]                                  #
#---------------------------------------------------------------------------------------------------------------------------------------#
#300.   Update log.                                                                                                                     #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   | Date |    20200522        | Version | 1.00        | Updater/Creator | Lu Robin Bin                                                #
#   |______|____________________|_________|_____________|_________________|_____________________________________________________________#
#   | Log  |Introduce the [eigenMapMatMult] method from [MatrixMultiplication.cpp] C++ function to replace [crossprod] and [tcrossprod] #
#   |      | functions from R base; which dramatically increase the efficiency by almost 20 times!                                      #
#   |      |Quote: https://stackoverflow.com/questions/35923787/fast-large-matrix-multiplication-in-r                                   #
#   |      |Note:                                                                                                                       #
#   |      | [1] You have to install the package [Rcpp] to enable dynamic compilation of [CPP] code                                     #
#   |      | [2] You also have to [Rcpp::sourceCpp] the dependent [CPP] code to introduce the C++ functions to R                        #
#   |      | [3] Use [Rfast::transpose] to transpose the corresponding matrix before multiplication in order to get the same result as  #
#   |      |      [crossprod] or [tcrossprod] respectively                                                                              #
#   |______|____________________________________________________________________________________________________________________________#
#   |___________________________________________________________________________________________________________________________________#
#   | Date |    20230114        | Version | 1.10        | Updater/Creator | Lu Robin Bin                                                #
#   |______|____________________|_________|_____________|_________________|_____________________________________________________________#
#   | Log  |[1] Introduce a function [match.arg.x] to enable matching args after mutation, e.g. case-insensitive match                  #
#   |______|____________________________________________________________________________________________________________________________#
#---------------------------------------------------------------------------------------------------------------------------------------#
#400.   User Manual.                                                                                                                    #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |See the [Full Test Program] section                                                                                                #
#---------------------------------------------------------------------------------------------------------------------------------------#
#500.   Dependent Facilities.                                                                                                           #
#---------------------------------------------------------------------------------------------------------------------------------------#
#   |100.   Dependent packages                                                                                                          #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |   |Rfast, Rcpp                                                                                                                    #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |300.   Dependent functions                                                                                                         #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#   |   |omniR$Stats                                                                                                                    #
#   |   |   |eigenMapMatMult       (See function definition in [MatrixMultiplication.cpp])                                              #
#   |   |-------------------------------------------------------------------------------------------------------------------------------#
#   |   |omniR$AdvOp                                                                                                                    #
#   |   |   |match.arg.x                                                                                                                #
#   |-----------------------------------------------------------------------------------------------------------------------------------#
#---------------------------------------------------------------------------------------------------------------------------------------#

#001. Append the list of required packages to the global environment
#Below expression is used for easy copy-paste from raw text strings instead of quoted ones.
lst_pkg <- deparse(substitute(c(
	Rfast, Rcpp
)))
#Quote: https://www.regular-expressions.info/posixbrackets.html?wlr=1
lst_pkg <- paste0(lst_pkg, collapse = '')
lst_pkg <- gsub('[[:space:]]', '', lst_pkg, perl = T)
lst_pkg <- gsub('^c\\((.+)\\)', '\\1', lst_pkg, perl = T)
lst_pkg <- unlist(strsplit(lst_pkg, ',', perl = T))
options( omniR.req.pkg = base::union(getOption('omniR.req.pkg'), lst_pkg) )

cov_matrix <- function(x,y=NULL,dim = c('col','row')){
	#010. Local parameters
	#[Quote: https://stackoverflow.com/questions/15595478/how-to-get-the-name-of-the-calling-function-inside-the-called-routine ]
	LfuncName <- deparse(sys.call()[[1]])
	#If above statement cannot find the name correctly, this function must have been called via [do.call] or else,
	# hence we need to traverse one layer above current one and extract the first argument of that call.
	if (grepl('^function.+$',LfuncName[[1]],perl = T)) LfuncName <- gsub('^.+?\\((.+?),.+$','\\1',deparse(sys.call(-1)),perl = T)[[1]]
	if (is.null(y)) y <- x
	y_pseudo <- identical(x, y)
	dim <- match.arg.x(dim, arg.func = tolower)

	#100. Transpose the matrices to facilitate a stable algorithm
	#It is preferred to handle transposition outside this function while always using [col] for calculation
	if (dim == 'row') {
		x <- Rfast::transpose(x)
		if (y_pseudo) {
			y <- x
		} else {
			y <- Rfast::transpose(y)
		}
	}
	n <- nrow(x)
	if (n != nrow(y)) stop('[',LfuncName,'][x] has different length of the other dimension to [y] when [dim==[',dim,']]!')

	#200. Clean the data
	#It is preferred to handle NAs outside this function
	#[Quote: https://stackoverflow.com/questions/8161836/how-do-i-replace-na-values-with-zeros-in-an-r-dataframe ]
	# x[is.na(x)] <- 0
	# y[is.na(y)] <- 0

	#500. Calculation
	#[Formula: (cov(Xi,Yi) = sum of ( (Xij - mean(Xi)) * (Yij - mean(Yi)) ) / (n-1) ]
	#510. Adjust the center of Xi by: [Xij - mean(Xi)]
	#20200522 We do not use [Rfast::colmeans] as it has no parameter of [na.rm]
	mns_x <- colMeans( x, na.rm = T )
	x <- x - Rfast::rep_row( mns_x, nrow(x) )
	if (y_pseudo) {
		y <- x
	} else {
		mns_y <- colMeans( y, na.rm = T )
		y <- y - Rfast::rep_row( mns_y, nrow(y) )
	}

	#590. Calculate the covariance for columns in pairs
	return( eigenMapMatMult( Rfast::transpose(x) , y ) / ( n - 1 ) )
}

#[Full Test Program;]
if (FALSE){
	#Simple test
	if (TRUE){
		tmcn::setchs(rev=F)
		omniR <- 'D:\\R\\omniR'
		#Below may take quite a time, almost 20s
		Rcpp::sourceCpp(normalizePath(file.path(omniR, 'Stats', 'MatrixMultiplication.cpp')))

		x <- Rfast::matrnorm(100000, 1000)
		y <- Rfast::matrnorm(100000, 1000)

		#[AMD FX-6300 6Core 3.5G]

		#Test speed against the official R function [cov()], 20 times faster!
		start_time <- Sys.time()
		cov_xy1 <- cov_matrix(x,y)
		end_time <- Sys.time()
		end_time - start_time
		#[RAM 2GB]
		#[CPU 100%] 9.32s

		start_time <- Sys.time()
		cov_xy2 <- cov(x,y)
		end_time <- Sys.time()
		end_time - start_time
		#No extra RAM required
		#[CPU 20%] 170s (2.82m)

		all.equal(cov_xy1,cov_xy2)

		#Test speed against the [Rfast] function [cova()], 10 times faster!
		start_time <- Sys.time()
		cov_x1 <- cov_matrix(x)
		end_time <- Sys.time()
		end_time - start_time
		#[RAM 2GB]
		#[CPU 100%] 9.15s

		start_time <- Sys.time()
		cov_x2 <- Rfast::cova(x)
		end_time <- Sys.time()
		end_time - start_time
		#[RAM 2GB]
		#[CPU 20%] 84s (1.4m)

		all.equal(cov_x1,cov_x2)

	}
}
